File: 알고리즘 스터디 문제들/Part 6 - DFS와 BFS/문제/미로 탐색.py
from collections import deque


def bfs(N, M, maze, visited):

    dx = [0, 0, -1, 1]
    dy = [1, -1, 0, 0]

    Q = deque()
    Q.appendleft([0, 0])
    visited[0][0] = 1

    while (Q):
        cur = Q.pop()
        if (cur == [N - 1, M - 1]):
            return

        for i in range(4):
            y = cur[0] + dy[i]
            x = cur[1] + dx[i]

            if (x >= M or x < 0):
                continue
            if (y >= N or y < 0):
                continue
            if (visited[y][x] == -1 and maze[y][x] == '1'):
                Q.appendleft([y, x])
                visited[y][x] = visited[cur[0]][cur[1]] + 1


N, M = map(int, input().split())
maze = [list(input()) for i in range(N)]
visited = [[-1 for j in range(M)] for i in range(N)]

bfs(N, M, maze, visited)
print(visited[N - 1][M - 1])

File: 알고리즘 스터디 문제들/Part 6 - DFS와 BFS/문제/미로 탐색.cpp
#include <iostream>
#include <array>
#include <queue>
using namespace std;

array<array<bool, 101>, 101> maze;
array<array<int, 101>, 101> visited;

array<int, 4> dx = {0, 0, -1, 1};
array<int, 4> dy = {1, -1, 0, 0};

void bfs(size_t N, size_t M)
{
    queue<pair<int, int>> Q;
    Q.push(make_pair(1, 1));
    visited[1][1] = 1;

    while (!Q.empty())
    {
        pair<int, int> cur = Q.front();
        Q.pop();

        if (cur.second == M && cur.first == N)
            return;

        for (size_t i = 0; i < 4; i++)
        {
            int x = cur.second + dx[i];
            int y = cur.first + dy[i];

            if (x > M || x <= 0)
                continue;
            if (y > N || y <= 0)
                continue;
            if (visited[y][x] == -1 && maze[y][x])
            {
                Q.push(make_pair(y, x));
                visited[y][x] = visited[cur.first][cur.second] + 1;
            }
        }
    }
}

int main()
{
    size_t N, M;
    cin >> N >> M;

    char input[101];
    for (size_t i = 1; i <= N; i++)
    {
        cin >> input;
        for (size_t j = 1; j <= M; j++)
            maze[i][j] = input[j - 1] - '0';
        visited[i].fill(-1);
    }

    bfs(N, M);
    cout << visited[N][M];
}